packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesController.java

```diff
@@ -0,0 +1,512 @@
+package expo.modules.updates;
+
+import android.content.Context;
+import android.content.pm.ApplicationInfo;
+import android.content.pm.PackageManager;
+import android.net.ConnectivityManager;
+import android.net.Uri;
+import android.os.AsyncTask;
+import android.os.Handler;
+import android.os.HandlerThread;
+import android.os.Looper;
+import android.util.Log;
+
+import com.facebook.react.ReactApplication;
+import com.facebook.react.ReactInstanceManager;
+import com.facebook.react.ReactNativeHost;
+import com.facebook.react.bridge.Arguments;
+import com.facebook.react.bridge.JSBundleLoader;
+import com.facebook.react.bridge.ReactContext;
+import com.facebook.react.bridge.WritableMap;
+import com.facebook.react.modules.core.DeviceEventManagerModule;
+
+import androidx.annotation.Nullable;
+import expo.modules.updates.db.Reaper;
+import expo.modules.updates.db.UpdatesDatabase;
+import expo.modules.updates.db.entity.AssetEntity;
+import expo.modules.updates.db.entity.UpdateEntity;
+import expo.modules.updates.launcher.EmergencyLauncher;
+import expo.modules.updates.launcher.Launcher;
+import expo.modules.updates.launcher.LauncherWithSelectionPolicy;
+import expo.modules.updates.launcher.SelectionPolicy;
+import expo.modules.updates.launcher.SelectionPolicyNewest;
+import expo.modules.updates.loader.EmbeddedLoader;
+import expo.modules.updates.manifest.Manifest;
+import expo.modules.updates.loader.RemoteLoader;
+
+import java.io.File;
+import java.lang.ref.WeakReference;
+import java.lang.reflect.Field;
+import java.util.Map;
+
+public class UpdatesController {
+
+  private static final String TAG = UpdatesController.class.getSimpleName();
+
+  private static final String UPDATES_EVENT_NAME = "Expo.nativeUpdatesEvent";
+  private static final String UPDATE_AVAILABLE_EVENT = "updateAvailable";
+  private static final String UPDATE_NO_UPDATE_AVAILABLE_EVENT = "noUpdateAvailable";
+  private static final String UPDATE_ERROR_EVENT = "error";
+
+  private static UpdatesController sInstance;
+
+  private WeakReference<ReactNativeHost> mReactNativeHost;
+
+  private Uri mManifestUrl;
+  private File mUpdatesDirectory;
+  private Exception mUpdatesDirectoryException;
+  private Launcher mLauncher;
+  private DatabaseHolder mDatabaseHolder;
+  private SelectionPolicy mSelectionPolicy;
+
+  // launch conditions
+  private boolean mIsReadyToLaunch = false;
+  private boolean mTimeoutFinished = false;
+  private boolean mHasLaunched = false;
+
+  private UpdatesController(Context context, Uri url) {
+    sInstance = this;
```

Comment:
For testing purposes it might be better to move this line into `initialize` so that a test can create an UpdatesController instance with fewer side effects